src/plugins/data/public/antlr/shared/utils.ts

```diff
@@ -104,49 +107,94 @@ export const fetchColumnValues = async (
   table: string,
   column: string,
   services: IDataPluginServices,
-  fieldInOsd: IndexPatternField | undefined,
+  indexPattern: IndexPattern,
   datasetType: string | undefined
 ): Promise<any[]> => {
-  if (!datasetType || !Object.values(DEFAULT_DATA.SET_TYPES).includes(datasetType)) {
-    return [];
-  }
+  const fieldInOsd = indexPattern.fields.find((f) => f.name === column);
 
-  // default to true/false values for type boolean
+  // For Boolean fields directly return the values
   if (fieldInOsd?.type === 'boolean') {
     return ['true', 'false'];
   }
 
-  const allowedType = ['string'];
-  // don't return values if ui settings prevent it or the field type isn't allowed
+  // Return cached Autocomplete Results if available
+  if (fieldInOsd?.spec.autoCompleteValues && fieldInOsd?.spec.autoCompleteValues.length > 0) {
+    return fieldInOsd.spec.autoCompleteValues;
+  }
+
+  // Return topQueryValues if available and fire async API call to update the cache for subsequent calls
   if (
-    !services.uiSettings.get(UI_SETTINGS.QUERY_ENHANCEMENTS_SUGGEST_VALUES) ||
-    !fieldInOsd ||
+    fieldInOsd?.spec.topQueryValues &&
+    fieldInOsd.spec.topQueryValues.length > 0 &&
     !allowedType.includes(fieldInOsd.type)
   ) {
-    return [];
+    // Fire async API call to update cache in non-blocking manner
+    updateFieldValuesAsync(table, column, services, indexPattern, datasetType, fieldInOsd);
+    return fieldInOsd.spec.topQueryValues;
+  }
+
+  // Fire a synchronous query to fetch values
+  await updateFieldValuesAsync(table, column, services, indexPattern, datasetType, fieldInOsd);
+
+  // Return the results of synchronous calls
+  return fieldInOsd?.spec?.autoCompleteValues ?? [];
+};
+
+// Non-blocking async function to update field values in background
+const updateFieldValuesAsync = async (
+  table: string,
+  column: string,
+  services: IDataPluginServices,
+  indexPattern: IndexPattern,
+  datasetType: string | undefined,
+  fieldInOsd: IndexPatternField | undefined
+): Promise<void> => {
+  try {
+    // Check if conditions allow API call
+    if (!datasetType || !Object.values(DEFAULT_DATA.SET_TYPES).includes(datasetType)) {
+      return;
+    }
+    if (
+      !services.uiSettings.get(UI_SETTINGS.QUERY_ENHANCEMENTS_SUGGEST_VALUES) ||
+      !fieldInOsd ||
+      !allowedType.includes(fieldInOsd.type) ||
+      !indexPattern
+    ) {
+      return;
+    }
+
+    const limit = services.uiSettings.get(UI_SETTINGS.QUERY_ENHANCEMENTS_SUGGEST_VALUES_LIMIT);
+    const dataset = services.data.query.queryString.getQuery().dataset;
```

Comment:
I think you should just use the index pattern passed in . The best option tho is if you have passed in the query object. At which point u access query.dataset to get the type of dataset and the full index pattern object from the service. Like directly passed in not from accessing a Singleton. To protect ourselves from race conditions